### PR TITLE
Complex paths

### PR DESCRIPTION
**Pull request by fidus58**
complex paths as new centrality algorithm

tests are still missing

### CODE REVIEW

File: networkit/centrality.pyx [hunk 1]
```diff
@@ -2928,3 +2932,120 @@ class SciPyPageRank(SpectralCentrality):
...
+		vector[double] getPLci() except +
+		void normalize() except +
+		vector[node] getAdopters() except +
+		_Graph &getComplexGraph() except +
+
+cdef class ComplexPaths(Algorithm):
+	""" 
+	ComplexPaths(G, threshold, mode, start)
+
+	according to:
```
Comment (fabratu) [inline networkit/centrality.pyx:2960]: Please elaborate more what complex paths does. Take a look at the documentation of other recently implemented algorithms for reference.

File: networkit/centrality.pyx [hunk 2]
```diff
@@ -2928,3 +2932,120 @@ class SciPyPageRank(SpectralCentrality):
...
+ 
+	Parameters
+	----------
+	G : networkit.Graph
+		The graph.
+
+	threshold : int
+		The complex path width: minimal number of neighbors
+
+	mode : ComplexPathMode
```
Comment (fabratu) [inline networkit/centrality.pyx:2974]: Writing the parameter this way will likely break the rendering of the doc. Take a look at other examples in NetworKit at how this is written - for example `Diameter`:  https://github.com/networkit/networkit/blob/master/networkit/distance.pyx#L487

Also, the overview over included enum-parameters should be added above the parameter section.

File: networkit/centrality.pyx [hunk 3]
```diff
@@ -2928,3 +2932,120 @@ class SciPyPageRank(SpectralCentrality):
...
+			mode=ComplexPathMode.singleNode
+		self._G = G
+		self._this = new _ComplexPaths(G._this, threshold, mode, start)
+
+	def getPLci(self):
+		""" 
+		getPLci()
+
+		Returns complex path lengths for every node in G
+		absolute or scaled to [0,1] when normalize() was called before run
```
Comment (fabratu) [inline networkit/centrality.pyx:2998]:
```
... before run.
This function is available only if networkit.centrality.ComplexPathMode.ALL_NODES was choosen.
```

File: networkit/centrality.pyx [hunk 4]
```diff
@@ -2928,3 +2932,120 @@ class SciPyPageRank(SpectralCentrality):
...
+
+cdef extern from "<networkit/centrality/ComplexPaths.hpp>" namespace "NetworKit::ComplexPathAlgorithm":
+
+	cpdef enum _Mode  "NetworKit::ComplexPathAlgorithm::Mode":
+		singleNode,
+		allNodes
+
+cdef class ComplexPathMode(object):
+	singleNode = _Mode.singleNode
+	allNodes   = _Mode.allNodes
```
Comment (fabratu) [inline networkit/centrality.pyx:2946]: Enums in Python are written in capital letters with underscore (SREAMING_SNAKE_CASE): `SINGLE_NODE` and `ALL_NODES`

File: networkit/centrality.pyx [hunk 5]
```diff
@@ -2928,3 +2932,120 @@ class SciPyPageRank(SpectralCentrality):
...
+		""" 
+		getPLci()
+
+		Returns complex path lengths for every node in G
+		absolute or scaled to [0,1] when normalize() was called before run
+		in mode ComplexPathMode.allNodes only
+
+		Returns
+		-------
+		list(double)
```
Comment (fabratu) [inline networkit/centrality.pyx:3003]: `float`

File: networkit/centrality.pyx [hunk 6]
```diff
@@ -2928,3 +2932,120 @@ class SciPyPageRank(SpectralCentrality):
...
+			A vector containing complex path lengths for all nodes
+		"""
+		return (<_ComplexPaths*>(self._this)).getPLci()
+
+	def getComplexGraph(self):
+		""" 
+		getComplexGraph()
+
+		Returns complex path (sub)graph of G starting in start
```
Comment (fabratu) [inline networkit/centrality.pyx:3012]: `in node start` (also use `:code` for `start`)

File: networkit/centrality.pyx [hunk 7]
```diff
@@ -2928,3 +2932,120 @@ class SciPyPageRank(SpectralCentrality):
...
+		in mode ComplexPathMode.singleNode only
```
Comment (fabratu) [inline networkit/centrality.pyx:3013]: New sentence

File: networkit/centrality.pyx [hunk 8]
```diff
@@ -2928,3 +2932,120 @@ class SciPyPageRank(SpectralCentrality):
...
+		"""
+		return Graph().setThis((<_ComplexPaths*>(self._this)).getComplexGraph())
+
+	def getAdopters(self):
+		"""
+		getAdopters()
+
+		Returns all nodes in the complex subgraph with at least
+		threshold neighbors (those who are adopted/infected when 
+		starting in start)
```
Comment (fabratu) [inline networkit/centrality.pyx:3027]: MIssing punctuation

File: networkit/centrality.pyx [hunk 9]
```diff
@@ -2928,3 +2932,120 @@ class SciPyPageRank(SpectralCentrality):
...
+
+		Returns
+		-------
+		list(node)
```
Comment (fabratu) [inline networkit/centrality.pyx:3031]: `int`

File: networkit/centrality.pyx [hunk 10]
```diff
@@ -2928,3 +2932,120 @@ class SciPyPageRank(SpectralCentrality):
...
+	def normalize(self):
+		"""
+		normalize()
+
+		when called before run() all complex path lengths 
+		returned by getPLci are scaled to [0,1]
+
+		in mode ComplexPathMode.allNodes only
+
+		Returns
```
Comment (fabratu) [inline networkit/centrality.pyx:3045]: Remove the `Returns` section from the docstring.

File: networkit/centrality.pyx [hunk 11]
```diff
@@ -2928,3 +2932,120 @@ class SciPyPageRank(SpectralCentrality):
...
+		list(node) 
+			A vector all adopted/infected nodes
+		"""
+		return (<_ComplexPaths*>(self._this)).getAdopters()
+
+	def normalize(self):
+		"""
+		normalize()
+
+		when called before run() all complex path lengths
```
Comment (fabratu) [inline networkit/centrality.pyx:3040]: `When called before run(), all ...`

File: networkit/centrality.pyx [hunk 12]
```diff
@@ -2928,3 +2932,120 @@ class SciPyPageRank(SpectralCentrality):
...
+			A vector all adopted/infected nodes
+		"""
+		return (<_ComplexPaths*>(self._this)).getAdopters()
+
+	def normalize(self):
+		"""
+		normalize()
+
+		when called before run() all complex path lengths 
+		returned by getPLci are scaled to [0,1]
```
Comment (fabratu) [inline networkit/centrality.pyx:3041]: `are scaled to [0,1].`

File: networkit/centrality.pyx [hunk 13]
```diff
@@ -2928,3 +2932,120 @@ class SciPyPageRank(SpectralCentrality):
...
+
+		in mode ComplexPathMode.allNodes only
```
Comment (fabratu) [inline networkit/centrality.pyx:3043]: No new line and convert it to a proper sentence.

File: networkit/centrality.pyx [hunk 14]
```diff
@@ -2928,3 +2932,120 @@ class SciPyPageRank(SpectralCentrality):
 
 		self.eigenvector = spectrum[1][0]
 		self.eigenvalue = spectrum[0][0]
+
+##### from branch ComplexPaths (K. Ahrens)
```
Comment (fabratu) [inline networkit/centrality.pyx:2936]: Remove this line (and unnecessary empty lines).

**Review by fabratu — CHANGES_REQUESTED** [review]

File: include/networkit/centrality/ComplexPaths.hpp [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+#define NETWORKIT_CENTRALITY_COMPLEX_PATHS_HPP_
+
+#include <random>
+#include <vector>
+
+#include <networkit/base/Algorithm.hpp>
+#include <networkit/graph/Graph.hpp>
+
+using std::sample;
+using std::vector;
```
Comment (angriman) [inline include/networkit/centrality/ComplexPaths.hpp:30]: I think we should drop this, it does not make the code easier to read.

File: networkit/cpp/centrality/ComplexPaths.cpp [hunk 16]
```diff
@@ -0,0 +1,203 @@
...
+}
+
+vector<double> ComplexPathAlgorithm::complexPathLength(count t) {
+    const auto &g = inputGraph;
+    auto n = g.numberOfNodes();
+    vector<double> PLci(n);
+    constexpr auto infDist = std::numeric_limits<edgeweight>::max();
+
+    g.parallelForNodes([&](node u) {
+        auto c_g = complexPathsGraph(u, t, nullptr);
```
Comment (angriman) [inline networkit/cpp/centrality/ComplexPaths.cpp:185]:
```suggestion
        Graph c_g = complexPathsGraph(u, t, nullptr);
```

File: networkit/cpp/centrality/ComplexPaths.cpp [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+        auto noPaths = false;
```
Comment (angriman) [inline networkit/cpp/centrality/ComplexPaths.cpp:186]:
```suggestion
        bool noPaths = false;
```

File: networkit/cpp/centrality/ComplexPaths.cpp [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+        NetworKit::BFS bfs(c_g, u, noPaths);
```
Comment (angriman) [inline networkit/cpp/centrality/ComplexPaths.cpp:187]: We are already in the `NetworKit` namespace.
```suggestion
        BFS bfs(c_g, u, noPaths);
```

File: networkit/cpp/centrality/ComplexPaths.cpp [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+        bfs.run();
+        auto distances = bfs.getDistances();
+        for (auto &d : distances)
+            if (d == infDist)
+                d = 0.0;
+        auto distanceSum = std::accumulate(distances.begin(), distances.end(), 0.0);
+        PLci[u] = distanceSum / n;
```
Comment (angriman) [inline networkit/cpp/centrality/ComplexPaths.cpp:194]: Division between `double` and `uint64_t`
```suggestion
        PLci[u] = distanceSum / static_cast<double>(n);
```

File: networkit/cpp/centrality/ComplexPaths.cpp [hunk 20]
```diff
@@ -0,0 +1,203 @@
...
+    g.parallelForNodes([&](node u) {
+        auto c_g = complexPathsGraph(u, t, nullptr);
+        auto noPaths = false;
+        NetworKit::BFS bfs(c_g, u, noPaths);
+        bfs.run();
+        auto distances = bfs.getDistances();
+        for (auto &d : distances)
+            if (d == infDist)
+                d = 0.0;
+        auto distanceSum = std::accumulate(distances.begin(), distances.end(), 0.0);
```
Comment (angriman) [inline networkit/cpp/centrality/ComplexPaths.cpp:193]: I think that we can do something like described in https://en.cppreference.com/w/cpp/algorithm/transform_reduce:
```suggestion
        const double distanceSum = std::transform_reduce(distances.cbegin(), distances.cend(), 0.0, std::plus{}, [](double val) { return val == infDist ? 0.0 : val; });
```

File: networkit/cpp/centrality/ComplexPaths.cpp [hunk 21]
```diff
@@ -0,0 +1,203 @@
...
+}
+
+static void addNewEdge(Graph &g, node u, node v) {
+    if (g.hasEdge(u, v))
+        return;
+    g.addEdge(u, v);
+}
+
+Graph ComplexPathAlgorithm::complexPathsGraph(node seed, count threshold, vector<node> *adopters) {
+    const auto &g = inputGraph;
```
Comment (angriman) [inline networkit/cpp/centrality/ComplexPaths.cpp:90]:
```suggestion
    const Graph &g = inputGraph;
```

File: networkit/cpp/centrality/ComplexPaths.cpp [hunk 22]
```diff
@@ -0,0 +1,203 @@
...
+            auto indirectNeighbors = Graph::NeighborRange<>(g, u);
+            for (auto in : indirectNeighbors) {
+                if (in != seed
+                    && std::find(moreSeeds.begin(), moreSeeds.end(), in) == moreSeeds.end())
+                    moreSeeds.push_back(in);
+            }
+            sample(moreSeeds.begin(), moreSeeds.end(), std::back_inserter(seeds), needMore,
+                   randomGen);
+            if (seeds.size() < needSeeds)
+                WARN("complexPathsGraph: ", seed, ": not enough indirect neighbors.");
```
Comment (angriman) [inline networkit/cpp/centrality/ComplexPaths.cpp:120]: This is inefficient and overly complicated. We basically need to sample a fixed number of seeds from the nodes up to distance 2 from the seed. Why don't we:
- Sample from the neighbors of `seed` only if they are enough, otherwise
- Run a BFS that stops at level 2 and use two queues to keep all the nodes at distance 1 and 2 separate. Then insert all the nodes at level 1 and sample the remaining number of seeds from the nodes at level 2.

In this way the code would be simpler and we won't need to run all these `std::find` (which are linear searches).

File: networkit/cpp/centrality/ComplexPaths.cpp [hunk 23]
```diff
@@ -0,0 +1,203 @@
...
+        }
+    }
+
+    seeds.insert(seeds.begin(), seed); // put in front, so that direct neighbors are activated first
+
+    for (auto u : seeds) {
```
Comment (angriman) [inline networkit/cpp/centrality/ComplexPaths.cpp:126]:
```suggestion
    for (node u : seeds) {
```

File: networkit/cpp/centrality/ComplexPaths.cpp [hunk 24]
```diff
@@ -0,0 +1,203 @@
...
+        activated[u] = true;
+        if (adopters)
+            adopters->push_back(u);
+        auto neighbors = Graph::NeighborRange<>(g, u);
+
+        for (auto v : neighbors) {
```
Comment (angriman) [inline networkit/cpp/centrality/ComplexPaths.cpp:132]:
```suggestion
        for (node v : neighbors) {
```

**Review by angriman — CHANGES_REQUESTED** [review]
Thanks for the new algorithm, I left some comments. My comments about the usage of `auto` are probably arguable, but as a reviewer I found some parts of the code hard to follow because of that. If using `auto` confuses the reader, then it was probably misused.

File: networkit/cpp/centrality/ComplexPaths.cpp [hunk 25]
```diff
@@ -0,0 +1,203 @@
...
+                    activated[u] = true;
+                    if (adopters)
+                        adopters->push_back(u);
+                    spread = true;
+                }
+                auto uNeighbors = Graph::NeighborRange<>(g, u);
+                for (auto v : uNeighbors) {
+                    addNewEdge(complex_g, u, v);
+                }
+            }
```
[the remark this answers is not in the document]
Comment (PetholzA) [reply networkit/cpp/centrality/ComplexPaths.cpp]: This results in the compiler warning: "control reaches the end of non-void function".

File: include/networkit/centrality/ComplexPaths.hpp [hunk 26]
```diff
@@ -0,0 +1,114 @@
...
+    Graph getComplexGraph();
+
+    /**
+     * nodes in the resulting graph after running in <code>Mode::singleNode</code>
+     * which are connected to start by at least threshold paths
+     */
+    vector<node> getAdopters();
+
+private:
+    const Graph &inputGraph;
```
[the remark this answers is not in the document]
Comment (fabratu) [reply include/networkit/centrality/ComplexPaths.hpp:96]: Can you elaborate more on what "these cases" means? 

I remember some discussion in the past about const ref vs. `const` pointer. However, I only see an usecase for pointer, if we don't initialize the member variable directly in the constructor.

**Review by fabratu — DISMISSED** [review]

**Comment by fabratu** [comment]
@angriman Thanks for the review - I want to merge this new algorithm into the next release. Any objections?

File: networkit/cpp/centrality/ComplexPaths.cpp (hunk 20 above)
Comment (angriman) [reply networkit/cpp/centrality/ComplexPaths.cpp:195]: Not resolved.

File: networkit/cpp/centrality/ComplexPaths.cpp (hunk 23 above)
Comment (angriman) [reply networkit/cpp/centrality/ComplexPaths.cpp:131]: Not resolved.

File: networkit/cpp/centrality/ComplexPaths.cpp [hunk 27]
```diff
@@ -0,0 +1,203 @@
...
+        });
+        if (spread) { // new influences in growing complex_g
+            complex_g.parallelForNodes(computeInfluence);
+        }
+    } while (spread);
+
+    return complex_g;
+}
+
+static auto min_max_norm(vector<double> data) {
```
[the remark this answers is not in the document]
Comment (angriman) [reply networkit/cpp/centrality/ComplexPaths.cpp]: Not resolged.

**Comment by angriman** [comment]
It seems to me that some comments were marked as resolved without resolving them nor a justification of why they were not resolved. I marked them as unresolved. If you disagree with some of them, please let me know.

File: networkit/cpp/centrality/ComplexPaths.cpp (hunk 20 above)
Comment (PetholzA) [reply networkit/cpp/centrality/ComplexPaths.cpp:195]: I did forgot to mention, that i tried using std::transform_reduce, this resulted in some older build on the CI failing (iirc: Linux - cpp8), because it did not knew std::transform_reduce yet.

File: networkit/cpp/centrality/ComplexPaths.cpp (hunk 25 above)
[the remark this answers is not in the document]
Comment (angriman) [reply networkit/cpp/centrality/ComplexPaths.cpp]: Maybe in my suggestion I forgot a parenthesis. Anyway, the idea is to avoid an unnecessary indentation level. Instead of wrapping everything around `if (influence[u] >= threshold)`, you can use `if (influence[u] < threshold) return;`, the rest of the lambda remains the same. That lambda returns `void` so that compiler error is unrelated the suggestion.

File: networkit/cpp/centrality/ComplexPaths.cpp (hunk 20 above)
Comment (angriman) [reply networkit/cpp/centrality/ComplexPaths.cpp:195]: Oh, pity.

File: networkit/cpp/centrality/ComplexPaths.cpp [hunk 28]
```diff
@@ -0,0 +1,205 @@
...
+        });
+        if (spread) { // new influences in growing complex_g
+            complex_g.parallelForNodes(computeInfluence);
+        }
+    } while (spread);
+
+    return complex_g;
+}
+
+static auto min_max_norm(std::vector<double> data) {
```
Comment (angriman) [inline networkit/cpp/centrality/ComplexPaths.cpp:168]:
```suggestion
static double min_max_norm(const std::vector<double> &data) {
```

File: networkit/cpp/centrality/ComplexPaths.cpp [hunk 29]
```diff
@@ -0,0 +1,205 @@
...
+    auto [min, max] = std::minmax_element(data.begin(), data.end());
+    double diff = *max - *min;
+    std::vector<double> norm(data);
+    // Hint: Use structured bind "min" as init-capture in order to avoid compiler errors
+    // for C++17 and Clang (at least up until C++20).
+    std::transform(norm.cbegin(), norm.cend(), norm.begin(),
+                   [diff, &min = min](auto e) { return (e - *min) / diff; });
```
Comment (angriman) [inline networkit/cpp/centrality/ComplexPaths.cpp:175]:
```suggestion
                   [diff, &min = min](double e) { return (e - *min) / diff; });
```

File: networkit/cpp/centrality/ComplexPaths.cpp [hunk 30]
```diff
@@ -0,0 +1,205 @@
...
+
+    std::vector<node> seeds;
+    seeds.reserve(4);
+
+    auto neighbors = Graph::NeighborRange<>(g, seed);
+
+    auto randomGen = Aux::Random::getURNG();
+    std::sample(neighbors.begin(), neighbors.end(), std::back_inserter(seeds), seedsNeeded,
+                randomGen);
+    auto needMore = seedsNeeded - seeds.size();
```
Comment (angriman) [inline networkit/cpp/centrality/ComplexPaths.cpp:99]: If `seeds.size()` is bigger than `seedsNeeded` this will overflow.
```suggestion
    if (seeds.size() >= seedsNeeded) return seeds;
    const count additionalSeeds = seedsNeeded - seeds.size();
```

File: networkit/cpp/centrality/ComplexPaths.cpp [hunk 31]
```diff
@@ -0,0 +1,205 @@
...
+#include <networkit/base/Algorithm.hpp>
+#include <networkit/centrality/ComplexPaths.hpp>
+#include <networkit/distance/BFS.hpp>
+#include <networkit/distance/PrunedLandmarkLabeling.hpp>
+#include <networkit/graph/Graph.hpp>
+
+namespace NetworKit {
+
+ComplexPathAlgorithm::ComplexPathAlgorithm(const Graph &G, count threshold, Mode mode, node start)
+    : inputGraph{&G}, mode{mode}, start{start}, threshold{threshold}, normPaths{false} {
```
Comment (angriman) [inline networkit/cpp/centrality/ComplexPaths.cpp:36]: Add a check for `threshold > 0`. Below in the code you have `auto seedsNeeded = threshold - 1` which will overflow if `threshold` is 0.

File: networkit/cpp/centrality/ComplexPaths.cpp [hunk 32]
```diff
@@ -0,0 +1,205 @@
...
+
+static void addNewEdge(Graph &g, node u, node v) {
+    if (g.hasEdge(u, v))
+        return;
+    g.addEdge(u, v);
+}
+
+std::vector<node> ComplexPathAlgorithm::generateSeeds(node seed, const Graph &g, count threshold) {
+    // activate threshold-1 random neighbors of seed:
+    auto seedsNeeded = threshold - 1;
```
Comment (angriman) [inline networkit/cpp/centrality/ComplexPaths.cpp:89]: Can we be 100% sure
```suggestion
    const count seedsNeeded = threshold - 1;
```

File: networkit/cpp/centrality/ComplexPaths.cpp [hunk 33]
```diff
@@ -0,0 +1,205 @@
...
+    auto needMore = seedsNeeded - seeds.size();
+    if (needMore > 0) {
+        WARN("complexPathsGraph: ", seed, ": not enough direct neighbors.");
+
+        // collect all nodes at distance 2 before sampling to establish equal distribution
+        std::vector<node> indirNeighbors;
+
+        for (node u : neighbors) {
+            auto indirectNeighbors = Graph::NeighborRange<>(g, u);
+            for (node in : indirectNeighbors) {
```
Comment (angriman) [inline networkit/cpp/centrality/ComplexPaths.cpp:108]:
```suggestion
            for (node in : g.neighborRange(u)) {
```

File: networkit/cpp/centrality/ComplexPaths.cpp [hunk 34]
```diff
@@ -0,0 +1,205 @@
...
+
+Graph ComplexPathAlgorithm::complexPathsGraph(node seed, count threshold,
+                                              std::vector<node> *adopters) {
+    const Graph *g = inputGraph;
+    const auto n = g->numberOfNodes();
+    Graph complex_g = Graph(n); // all nodes, no edges so far
+    std::vector<bool> activated(n);
+    std::vector<count> influence(n);
+
+    auto seeds = generateSeeds(seed, *g, threshold);
```
Comment (angriman) [inline networkit/cpp/centrality/ComplexPaths.cpp:128]:
```suggestion
    std::vector<node> seeds = generateSeeds(seed, *g, threshold);
```

File: networkit/cpp/centrality/ComplexPaths.cpp [hunk 35]
```diff
@@ -0,0 +1,205 @@
...
+            }
+        }
+        std::sample(indirNeighbors.begin(), indirNeighbors.end(), std::back_inserter(seeds),
+                    needMore, randomGen);
+    }
+    return seeds;
+}
+
+Graph ComplexPathAlgorithm::complexPathsGraph(node seed, count threshold,
+                                              std::vector<node> *adopters) {
```
Comment (angriman) [inline networkit/cpp/centrality/ComplexPaths.cpp:121]: Why don't we use a reference?
```suggestion
                                              std::vector<node> &adopters) {
```

**Review by angriman — CHANGES_REQUESTED** [review]
Thanks, I left a few more comments.

File: networkit/cpp/centrality/ComplexPaths.cpp (hunk 35 above)
Comment (PetholzA) [reply networkit/cpp/centrality/ComplexPaths.cpp:125]: As far as i can tell, because of line 191, where we call complexPathGraphs using the nullpointer. (is there a way around this?)

Comment (angriman) [reply networkit/cpp/centrality/ComplexPaths.cpp:126]: I see. Then let's keep the pointer.

**Review by angriman — APPROVED** [review]
Thanks, LGTM.
